src/middlewares/index.ts

```diff
@@ -0,0 +1,89 @@
+import jwt from 'jsonwebtoken'
+import { Response, Request, NextFunction } from 'express'
+import {  getUserById } from '../services/userServices'
+
+declare global {
+  // eslint-disable-next-line @typescript-eslint/no-namespace
+  namespace Express {
+    // eslint-disable-next-line no-shadow
+    interface Request {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      userDetails: any
+    }
+  }
+}
+
+//   removing token and extracting token data from the request headers
+
+export const extractToken = async (
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  try {
+    if (!req.headers.authorization) {
+        return res.status(401).json({  message: 'Please Login' });
+      }
+    const token = req.headers.authorization?.split(' ')[1] || ''
+
+    if (!token) {
+      return res.status(401).json({
+        message: 'no access token found',
+      })
+    }
+
+    jwt.verify(
+      token,
+      process.env.JWT_SECRET || 'USER-AUTH',
+      (err: unknown, decoded: unknown) => {
+        if (err) {
+          return res.status(401).json({ message: 'Invalid token' })
+        }
+        res.locals.decoded = decoded
+        req.userDetails = decoded
+        next()
+      },
+    )
+  } catch (error) {
+    res.status(500).json({ message: 'Internal server down' })
+  }
+}
+
+export const checkPermission =
+  (permissionRole: string) =>
+  async (req: Request, res: Response, next: NextFunction) => {
+    const userInfo = req.userDetails
+
+    const existingUser = await getUserById(userInfo?.id)
+    if (existingUser && existingUser?.userRole === permissionRole) {
+      return next()
+    }
+    return res.status(401).json({ code: 401, message: 'Unauthorized' })
+  }
+
+
+  export const isOwner =  async (req: Request, res: Response, next: NextFunction)=>{
+    try{
+        const {id} = req.params;
+        const userInfo = req.userDetails
+
+        if(!userInfo){
+            return res.status(401).json({message:"Unauthorized"});
+        }
+        
+        const existingUser = await getUserById(userInfo?.id)
+
+        if(!existingUser){
+            return res.status(401).json({message:"Unauthorized"});
+        }
+        if(existingUser.id as number !== Number(id)){
+            return res.status(401).json({message:"Unauthorized"})
+        }
+            next()
+        
```

Comment:
```suggestion

        
```
this have entirely new feature, this middleware doesn't have be included in this fix